kube-runtime/src/utils/predicate.rs

```diff
@@ -152,13 +182,29 @@ where
 
     fn poll_next(self: Pin<&mut Self>, cx: &mut Context<'_>) -> Poll<Option<Self::Item>> {
         let mut me = self.project();
+
+        // Evict expired entries before processing new events
+        let now = Instant::now();
+        let ttl = me.config.ttl;
+        me.cache
+            .retain(|_, entry| now.duration_since(entry.last_seen) < ttl);
+
         Poll::Ready(loop {
             break match ready!(me.stream.as_mut().poll_next(cx)) {
                 Some(Ok(obj)) => {
                     if let Some(val) = me.predicate.hash_property(&obj) {
                         let key = PredicateCacheKey::from(obj.meta());
-                        let changed = me.cache.get(&key) != Some(&val);
-                        me.cache.insert(key, val);
+                        let now = Instant::now();
+
+                        // Check if the predicate value changed or entry doesn't exist
+                        let changed = me.cache.get(&key).map(|entry| entry.hash) != Some(val);
```

Comment:
This can be slightly simplified with an [`Option::is_some_and`](https://doc.rust-lang.org/std/option/enum.Option.html#method.is_some_and).